packages/back-end/src/models/OrganizationModel.ts

```diff
@@ -434,3 +440,106 @@ export async function updateMember(
     }),
   });
 }
+
+export const customRoleValidator = z
+  .object({
+    id: z.string().min(2).max(64),
+    description: z.string().max(100),
+    policies: z.array(z.enum(POLICIES)),
+  })
+  .strict();
+
+export async function addCustomRole(org: OrganizationInterface, role: Role) {
+  // Basic Validation
+  role = customRoleValidator.parse(role);
+
+  // Make sure role id is not reserved
+  if (RESERVED_ROLE_IDS.includes(role.id)) {
+    throw new Error("That role id is reserved and cannot be used");
+  }
+
+  // Make sure role id is not already in use
+  if (org.customRoles?.find((r) => r.id === role.id)) {
+    throw new Error("That role id already exists");
+  }
+
+  // Validate custom role id format
+  if (!/^[a-zA-Z0-9_]+$/.test(role.id)) {
+    throw new Error(
+      "Role id must only include letters, numbers, and underscores."
+    );
+  }
+
+  const customRoles = [...(org.customRoles || [])];
+  customRoles.push(role);
+
+  await updateOrganization(org.id, { customRoles });
+}
+
+export async function editCustomRole(
+  org: OrganizationInterface,
+  id: string,
+  updates: Omit<Role, "id">
+) {
+  // Validation
+  updates = customRoleValidator.omit({ id: true }).parse(updates);
+
+  let found = false;
+  const newCustomRoles = (org.customRoles || []).map((role) => {
+    if (role.id === id) {
+      found = true;
+      return {
+        ...role,
+        ...updates,
+      };
+    }
+    return role;
+  });
+
+  if (!found) {
+    throw new Error("Role not found");
+  }
+
+  await updateOrganization(org.id, { customRoles: newCustomRoles });
+}
+
+function usingRole(member: MemberRoleWithProjects, role: string): boolean {
+  return (
+    member.role === role ||
+    (member.projectRoles || []).some((pr) => pr.role === role)
+  );
+}
+
+export async function removeCustomRole(
+  org: OrganizationInterface,
+  teams: TeamInterface[],
+  id: string
+) {
```

Comment:
I think we'll want to add one more check - to prevent deleting a role that is the org's default.
```ts
  // Make sure the id isn't the org's default
  if (org.settings?.defaultRole?.role === id) {
    throw new Error(
      "Cannot delete role. This role is set as the organization's default role."
    );
  }
  ```